homeassistant/components/landisgyr_heat_meter/sensor.py

```diff
@@ -37,194 +40,255 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+@dataclass
+class HeatMeterSensorEntityDescription(SensorEntityDescription):
+    """Heat Meter sensor description."""
+
+    value_fn: Callable | None = None
```

Comment:
```suggestion
    value_fn: Callable[[StateType | datetime], StateType | datetime] = lambda value: value
```
You could remove the repeated `value_fn=lambda value: value,` in your description then and also don't need to unwrap it when calling it (for mypy).

Reply:
Looks good. However, if I do this, mypy gives me the following error:
`error: Argument "value_fn" to "HeatMeterSensorEntityDescription" has incompatible type "Callable[[datetime], datetime]"; expected "Callable[[Union[Union[str, int, float, None], datetime]], Union[Union[str, int, float, None], datetime]]"  [arg-type]`

If I try the Union syntax, mypy also doesn't like it and suggests the "alternative" syntax. 
Any suggestion on how to do the properly?

Reply:
I've used the default, but not the typing because of the mismatch with dt_util.as_utc.

Reply:
Since you are using `getattr(HeatMeterResponse, response_key, None)` now, isn't that still a type mismatch for the default `None`?

Maybe passing the response dataclass would solve that
```py
value_fn: Callable[[HeatMeterResponse], StateType | datetime]
```
In the descriptions you could use that like
```py
value_fn = lambda res: res.power_max_kw
```
This would also imply and typecheck `response_key`.

Reply:
This seems to work fine. And I like that more is now done inside the descriptions. Although it would now require all attributes to be present in the response, or an exception is thrown. I'm not sure I like that, again with tight coupling in mind.

Reply:
Since the response is a Dataclass, it should be well defined which attributes are present, isn't it? I haven't exactly looked up how your used library handles that.

Besides that I guess `try: except:` is a very clean and pythonic way to handle such things (and with Python 3.11 very lightweight/fast too). I'd even use that in every case instead of `getattr`.

Reply:
Thanks again, I think have found a way that is clean and effective. Curious if you agree.

Reply:
Wait another second, I'll get rid of the getattr as well

Reply:
`response_key` isn't used now.

Reply:
Argh, true. I've moved everything inside the class making it obsolete. I'll have another look, tnx

Reply:
I've removed the `reponse_key `(the loose coupling is now in the `value_fn`). I've kept the getattr, since I'll have to make a seperate get-function otherwise (with try/catch) which is essentially the same.

Reply:
That could be done in `native_value()`.
I'm not sure what the upsides of loose coupling are in this case.